spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java

```diff
@@ -0,0 +1,212 @@
+	 * identified by {@link #entityClass} instead of a List.
+	 *
+	 * @param expectSingleResult true if a single result is expected.
+	 */
+	public void setExpectSingleResult(boolean expectSingleResult) {
+		this.expectSingleResult = expectSingleResult;
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "r2dbc:reactive-inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		TypeLocator typeLocator = this.evaluationContext.getTypeLocator();
+		if (typeLocator instanceof StandardTypeLocator) {
+			/*
+			 * Register the R2dbc Query DSL package so they don't need a FQCN for QueryBuilder, for example.
+			 */
+			((StandardTypeLocator) typeLocator).registerImport("org.springframework.data.relational.core.query");
+		}
+		this.initialized = true;
+	}
+
+	/**
+	 * Execute a {@link Query} returning its results as the Message payload.
+	 * The payload can be either {@link reactor.core.publisher.Flux} or
+	 * {@link reactor.core.publisher.Mono} of objects of type identified by {@link #entityClass},
+	 * or a single element of type identified by {@link #entityClass}
+	 * based on the value of {@link #expectSingleResult} attribute which defaults to 'false' resulting
+	 * {@link org.springframework.messaging.Message} with payload of type
+	 * {@link reactor.core.publisher.Flux}. The collection name used in the
+	 */
+	@Override
+	public Object doReceive() {
+		Assert.isTrue(this.initialized, "This class is not yet initialized. Invoke its afterPropertiesSet() method");
+		Assert.notNull(this.queryExpression, "'queryExpression' must not be null");
+		Object value = this.queryExpression.getValue(this.evaluationContext);
+		Object result;
+		if (value instanceof Query) {
+			result = executeQuery(((Query) value));
+		}
+		else if (value instanceof String) {
+			result = executeQuery((String) value);
+		}
+		else {
+			throw new IllegalStateException("'queryExpression' must evaluate to String " +
+					"or org.springframework.data.relational.core.query.Query, but not: " + value);
+		}
+		return getMessageBuilderFactory()
+				.withPayload(result);
+	}
+
+	private Object executeQuery(String queryString) {
+		Object result;
+		if (this.expectSingleResult) {
+			result = this.r2dbcEntityOperations.getDatabaseClient()
+					.execute(queryString)
+					.as(this.entityClass)
+					.fetch()
+					.one();
+		}
+		else {
+			result = this.r2dbcEntityOperations.getDatabaseClient()
+					.execute(queryString)
+					.as(this.entityClass)
+					.fetch()
+					.all();
+		}
+		return result;
+	}
+
+	private Object executeQuery(Query query) {
+		Object result;
+		if (this.expectSingleResult) {
+			result = this.r2dbcEntityOperations.selectOne(query, this.entityClass);
```

Comment:
I think we should not rely on the `@Table` on the provided class.
I really believe that in most cases people are going to deal with just a `HashMap`.
Just because we interact here with messaging system and it is really hard to carry a particular domain class from one endpoint to another. 

So, I think we should get rid of `r2dbcEntityOperations` contract over here and just deal with the plain `DatabaseClient`. This way end-user will be able to not tie his/her logic with persistence model and those extra annotations.

We probably need to revise the `Query`  logic then in favor of `SelectSpec`...